quinn-proto/src/connection/mod.rs

```diff
@@ -2645,7 +2760,13 @@ where
             .push_back(EndpointEventInner::NeedIdentifiers(n));
     }
 
-    fn populate_packet(&mut self, space_id: SpaceId, buf: &mut Vec<u8>) -> SentFrames {
+    fn populate_packet(
+        &mut self,
+        space_id: SpaceId,
+        buf: &mut Vec<u8>,
+        buf_capacity: usize,
+        has_more: &mut bool,
```

Comment:
Rather than introducing this parameter, could we rerun the logic `poll_transmit` uses in `SpaceId::iter().filter(...)` to determine whether a space is ready to send as many times as necessary? This should give more precise, consistent results, and cause less churn.

Reply:
I don't think it's more precise. The false positives (like the reset stream) seem to happen in both cases. Maybe those should all be removed in separate cleanup efforts. I'm not sure whether false wakeups also lead to other unexpected side-effects (like sending more ACKs, which happened with the MAX_DATA frames).

But I agree using the same mechanism would be more streamlined.

Reply:
You're correct that there are still false positives due to reset streams (perhaps we could correct these cheaply by keeping a counter of reset streams that were previously pending, and comparing it to the length of the pending list?). However, I think `has_more` introduces false *negatives*, such as ignoring the application datagram queue. These could be fixed, but would represent a duplication of existing logic.